README.md
# mx2mt

*Limited MusicXML to MusixTeX converter*

This converter is primary used to transfer songs of [Falešné společenstvo](http://fs.ulmus.cz/) from [Mozart](http://www.mozart.co.uk/) music processor to [MusixTeX](https://www.ctan.org/pkg/musixtex) typesetting notation via small subset of [MusicXML](http://www.musicxml.com/) 3.0 format.

Written in [PHP](http://php.net/), sorry. But you can try [online version](http://mx2mt.ulmus.cz/). PHP 8 is not suported.

You also need small [header.tex](https://github.com/jirka-grunt/mx2mt/blob/master/header.tex) file for some auxiliary and compatibility stuff.

Supported features:
- score-partwise and score-timewise XML
- clefs: treble
- meters: with single denominator (C symbols for 2/2 and 4/4)
- keys: flats or sharps
- ordinary notes and rests
- lengths: from 1/32 to whole note, breve and longa
- heights: from C2 to H6 (5 octaves)
- accidentals: sharp, double sharp, natural, flat, double flat
- single dots from 1/16 notes and rests
- tuplets: triplet and duplet
- articulations: accent, staccato, fermata
- beams: single, duble, partial, dotted
- multiple measures lasting rests
- slurs and ties
- repeats: left, right, both
- endings: begin and end volta
- double bar lines
- chord (non-spacing) notes
- multiple staves
- guitar chords: major/minor, augmented/diminished, sevenths (with forenamed only), suspended and power (perfect fifth)

Some features may be added or extended in future (but most of them probably not).

Additional features:
- correct vertical alignment with multiple staves
- merging meters/signatures to general ones when possible
- automatic orientation of stems, beams, slurs and ties

Known limitations:
- only beams non-overlapping within one staff are supported
- bar lines & voltas are synchronized over all staves (and taken from first staff)
- number of staves can't be changed
- fermatas and tuplets have fixed height
- automatic orientation of stems, beams, slurs and ties
- guitar chords are in Czech (=German) notation: B -> H, Bb -> B
- augmented, diminished and half diminished seventh is not supported in chords
- wrong vertical alignment with tuplets in multiple staves

As MusicXML is pretty complex format there definitely are plenty of unknown limitations.
Moreover, conversion is tested only with XML exported from Mozart version 12.

Note: You can also give a chance to xml2pmx tool (binary distribution only).
It convers MusicXML to PMX format, which is pre-processor for MusixTeX. See
[archive](https://icking-music-archive.org/software/htdocs/) for details.


CHANGES
Version 1.2
===========

New features
------------

- Basic support for guitar chords: major and minor, augmented and diminished,
  sevens (with forementioned only), suspended (2nd and 4th) and power (perfect
  fifth).

- Meters 2/2 and 4/4 are replaced with traditional C symbols.

- Basic tuplets (triplets and duplets) are supported in standard situations.

- Breve (double whole) and longa (double breve) note lengths are supported.

Bugs fixed
----------

- Better default page layout handling.

- Repeats at the very beginning and the very end of piece are handled correctly.

- Fixed handling of unsupported features with multiple staves.

Version 1.1
===========

New features
------------

- Accidentals: sharp, double sharp, natural, flat, double flat.

- Chord (non-spacing) notes.

- 32nd notes and rests are now supported.

- Articulations: accent, staccato and fermata.

- Multiple measures lasting rests are now supported.

Bugs fixed
----------

- Better bar lines processing. Left and right bar lines from XML are
  merged to catch both voltas and repeats.

- Measures longer than whole note (like 12/8) are now supported.

- PHP 5.5+ is supported (Generator is now language internal class)

- Better processing of some unsupported features.

- Function for numbering slurs/ties now works with multiple stuffs.

- Spacing for pointed whole note fixed (MusiXTeX has no such spacing,
  so version for regular whole note is used).

- Voltas different from single digit are supported.



php-source/data.php
<?php

	// ----- from parser -----

	class Music {
		public $measures = array();
	}

	class Measure {
		public $parts = array();
	}

	class Part {
		public $attributes = array();
		public $durations = array();
		public $endings = array();
	}

	class Unsupported {
		public $info;
	}

	// ----- for generator -----

	class Notation {
		public $elements = array();
	}

	class Notes {
		public $long;
		public $parts = array();
	}

	class Nil {
	}

	// ----- ends -----

	abstract class Ending {
	}

	class Volta extends Ending {
		public $start;
		public $number;
	}

	class Repeat extends Ending {
		public $left;
		public $right;
	}

	class DoubleBar extends Ending {
	}

	// ----- divisions -----

	abstract class Division {
		public $number;
		public $leftRepeat;
		public $rightRepeat;
		public $double;
	}

	class StartPiece extends Division {
		public $parts;
		public $attributes = array();
	}

	class Bar extends Division {
	}

	class ChangeContext extends Division {
		public $attributes = array();
	}

	class EndPiece extends Division {
	}

	// ----- attributes -----

	abstract class Attribute {
	}

	class Meter extends Attribute {
		public $beats;
		public $unit;
	}

	class Signature extends Attribute {
		public $fifths;
	}

	// ----- durations -----

	abstract class Duration {
		public $long;
		public $connections = array();
		public $fermata;
		public $gchord;
	}

	class Pause extends Duration {
	}

	class FullPause extends Duration {
		public $count;
	}

	class Note extends Duration {
		public $height;
		public $alter;
		public $nolyr;
		public $chords = array();
		public $articulations = array();
		public $tuplet;
	}


	abstract class Beam extends Note {
		public $number;
		public $up;
		public $multiplicity;
		public $partial;
	}

	class BeamStart extends Beam {
		public $slant;
	}

	class BeamContinue extends Beam {
		public $change;
	}

	class BeamEnd extends Beam {
	}

	// ----- connections -----

	abstract class Connection {
		public $number;
	}

	abstract class ConnectionStart extends Connection {
		public $up;
	}

	abstract class ConnectionEnd extends Connection {
	}

	class TieStart extends ConnectionStart {
	}

	class TieEnd extends ConnectionEnd {
	}

	class SlurStart extends ConnectionStart {
	}

	class SlurEnd extends ConnectionEnd {
	}

	// ----- articulations -----

	abstract class Articulation {
		public $below;
	}

	class Accent extends Articulation {
	}

	class Staccato extends Articulation {
	}

	// ----- guitar -----

	class Chord {
		public $base;
		public $unsupported;
		public $minor = FALSE;
		public $seven = FALSE;
		public $alter = 0;
		public $plus = 0;
		public $maj = FALSE;
		public $sus = 0;
		public $power = FALSE;
	}



php-source/creator.php
<?php

	class Creator {

		protected $text = '';

		private $pitches = array();

		protected $notes = array(
				1   => '\notes',
				2   => '\notes',
				3   => '\notesp',
				4   => '\Notes',
				6   => '\Notesp',
				8   => '\NOtes',
				12  => '\NOtesp',
				16  => '\NOTes',
				24  => '\NOTesp',
				32  => '\NOTEs',
				48  => '\NOTEs',
				64  => '\NOTEs',
				96  => '\NOTEs',
				128 => '\NOTEs',
				192 => '\NOTEs',
			);

		protected $note = array(
				1   => '\ccca',
				2   => '\cca',
				3   => '\ccap',
				4   => '\ca',
				6   => '\cap',
				8   => '\qa',
				12  => '\qap',
				16  => '\ha',
				24  => '\hap',
				32  => '\wh',
				48  => '\whp',
				64  => '\wq',
				96  => '\wqp',
				128 => '\wqq',
				192 => '\wqqp',
			);

		protected $znote = array(
				1   => '\zq',
				2   => '\zq',
				3   => '\zqp',
				4   => '\zq',
				6   => '\zqp',
				8   => '\zq',
				12  => '\zqp',
				16  => '\zh',
				24  => '\zhp',
				32  => '\zwh',
				48  => '\zwhp',
				64  => '\zwq',
				96  => '\zwqp',
				128 => '\zwqq',
				192 => '\zwqqp',
			);

		protected $pause = array(
				1  => '\hs',
				2  => '\qs',
				3  => '\qsp',
				4  => '\ds',
				6  => '\dsp',
				8  => '\qp',
				12 => '\qpp',
				16 => '\hpause',
				24 => '\hpausep',
				32 => '\pause',
				48 => '\pausep',
			);


		public function __construct() {
			$upper = ord('C');
			$lower = ord('a');
			for($i=1; $i<=35; $i++) {
				if ($i <= 12) {
					$this->pitches[$i] = chr($upper);
					$upper++;
				} else {
					$this->pitches[$i] = chr($lower);
					$lower++;
				}
			}
		}

		public function getTex(Notation $notation) {
			$this->text = '';
			$this->writeBegin();
			foreach ($notation->elements as $element) {
				$name = get_class($element);
				switch ($name) {
					case 'Notes':
						$this->writeNotes($element);
						break;
					case 'Volta':
						$this->writeVolta($element);
						break;
					case 'StartPiece':
						$this->writeStartPiece($element);
						break;
					case 'Bar':
						$this->writeBar($element);
						break;
					case 'ChangeContext':
						$this->writeChangeContext($element);
						break;
					case 'EndPiece':
						$this->writeEndPiece($element);
						break;
					case 'Unsupported':
						$this->writeUnsupported($element);
						break;
				}
			}
			$this->writeEnd();
			return $this->text;
		}


		protected function writeBegin() {
			$this->add('\input header');
			$this->add('');
		}

		protected function writeEnd() {
			$this->add('');
			$this->add('\bye');
		}

		protected function writeNotes(Notes $notes) {
			$line = $this->notes[$notes->long];
			$parts = array();
			$count = count($notes->parts);
			$i = 1;
			foreach ($notes->parts as $part) {
				$name = get_class($part);
				$treat = TRUE;
				switch ($name) {
					case 'Pause':
						$fragment = $this->getPause($part);
						$treat = FALSE;
						break;
					case 'FullPause':
						$fragment = $this->getFullPause($part);
						$treat = FALSE;
						break;
					case 'Note':
						$fragment = $this->getNote($part);
						break;
					case 'BeamStart':
						$fragment = $this->getBeamStart($part);
						break;
					case 'BeamContinue':
						$fragment = $this->getBeamContinue($part);
						break;
					case 'BeamEnd':
						$fragment = $this->getBeamEnd($part);
						break;
					case 'Nil':
						$treat = FALSE;
						$fragment = '';
				}
				if ($treat && ($i != $count)) {
					$fragment .= '%';
				}
				if ($part instanceof Duration) {
					$fragment = $this->getGchord($part->gchord).$fragment;
				}
				$parts[] = $fragment;
				$i++;
			}
			$line .= implode("\n\t&", $parts);
			$line .= '\en';
			$this->add($line);
		}

		protected function writeVolta(Volta $volta) {
			if ($volta->start) {
				$this->add('\Setvolta'.$this->nr($volta->number).'%');
			} else {
				$this->add('\setendvolta');
			}
		}

		protected function writeStartPiece(StartPiece $start) {
			$this->add('\setvoices{'.$start->parts.'}');
			$this->writeAttributes($start->attributes);
			$this->add('');
			$this->add('\Startpiece');
			$this->add('% '.$start->number);
			if ($start->leftRepeat) {
				$this->add('\zleftrepeat');
			}
		}

		protected function writeBar(Bar $bar) {
			if ($bar->leftRepeat && $bar->rightRepeat) {
				$this->add('\leftrightrepeat');
			} elseif ($bar->leftRepeat) {
				$this->add('\leftrepeat');
			} elseif ($bar->rightRepeat) {
				$this->add('\rightrepeat');
			} elseif ($bar->double) {
				$this->add('\doublebar');
			} else {
				$this->add('\bar');
			}
			$this->add('% '.$bar->number);
		}

		protected function writeChangeContext(ChangeContext $context) {
			if ($context->leftRepeat && $context->rightRepeat) {
				$this->add('\setleftrightrepeat');
			} elseif ($context->leftRepeat) {
				$this->add('\setleftrepeat');
			} elseif ($context->rightRepeat) {
				$this->add('\setrightrepeat');
			} elseif ($context->double) {
				$this->add('\setdoublebar');
			}
			$this->writeAttributes($context->attributes);
			$this->add('\changecontext');
			$this->add('% '.$context->number);
		}

		protected function writeEndPiece(EndPiece $end) {
			if ($end->rightRepeat) {
				$this->add('\setrightrepeat');
				$this->add('\endpiece');
			} else {
				$this->add('\Endpiece');
			}
		}

		protected function writeAttributes(array $parts) {
			$singleMeter = TRUE;
			$singleSignature = TRUE;
			$firstMeter = NULL;
			$firstSignature = NULL;
			foreach ($parts as $index => $part) {
				$meter = NULL;
				$signature = NULL;
				foreach ($part as $attribute) {
					$name = get_class($attribute);
					switch ($name) {
						case 'Meter':
							if ($index == 0) {
								$firstMeter = $attribute;
							}
							$meter = $attribute;
							break;
						case 'Signature':
							if ($index == 0) {
								$firstSignature = $attribute;
							}
							$signature = $attribute;
							break;
					}
				}
				$singleMeter = ($singleMeter && ($meter == $firstMeter));
				$singleSignature = ($singleSignature && ($signature == $firstSignature));
			}

			if ($singleMeter) {
				if (!is_null($firstMeter)) {
					$this->writeMeter(0, $firstMeter);
				}
			} else {
				foreach ($parts as $index => $part) {
					foreach ($part as $attribute) {
						if ($attribute instanceof Meter) {
							$this->writeMeter($index+1, $attribute);
						}
					}
				}
			}
			if ($singleSignature) {
				if (!is_null($firstSignature)) {
					$this->writeSignature(0, $firstSignature);
				}
			} else {
				foreach ($parts as $index => $part) {
					foreach ($part as $attribute) {
						if ($attribute instanceof Signature) {
							$this->writeSignature($index+1, $attribute);
						}
					}
				}
			}
		}

		protected function writeMeter($part, Meter $meter) {
			if (($meter->beats == 2) && ($meter->unit == 2)) {
				$type = '\allabreve';
			} elseif (($meter->beats == 4) && ($meter->unit == 4)) {
				$type = '\meterC';
			} else {
				$type = '{\meterfrac{'.$meter->beats.'}{'.$meter->unit.'}}';
			}
			if ($part == 0) {
				$this->add('\generalmeter'.$type.'%');
			} else {
				$this->add('\setmeter'.$this->nr($part).'{'.$type.'}%');
			}
		}

		protected function writeSignature($part, Signature $signature) {
			if ($part == 0) {
				$this->add('\generalsignature{'.$signature->fifths.'}%');
			} else {
				$this->add('\setsign'.$this->nr($part).'{'.$signature->fifths.'}%');
			}
		}

		protected function writeUnsupported(Unsupported $unsupported) {
			$this->add('% ===== XXX: '.$unsupported->info.' =====');
		}


		protected function getPause(Pause $pause) {
			$fe = $this->getFermata($pause);
			$ps = $this->pause[$pause->long];
			return $fe.$ps;
		}

		protected function getFullPause(FullPause $pause) {
			$nr = '\nr'.$this->nr($pause->count);
			if ($pause->count === 2) {
				$ps = '\PAuse';
			} elseif ($pause->count === 4) {
				$ps = '\PAUSe';
			} else {
				// TODO
				$ps = '\pause';
			}
			return $nr.$ps;
		}

		protected function getNote(Note $note) {
			$nl = $this->nl($note->nolyr);
			$conn = $this->getConnections($note);
			$art = $this->getArticulation($note);
			$fe = $this->getFermata($note);
			$tu = $this->getTuplet($note);
			$lng = $this->note[$note->long];
			$hta = $this->hta($note, ' ');
			$znotes = $this->getZNotes($note);
			return $nl.$conn.$art.$fe.$tu.$znotes.$lng.$hta;
		}

		protected function getBeamStart(BeamStart $beam) {
			$nl = $this->nl($beam->nolyr);
			$conn = $this->getConnections($beam);
			$art = $this->getArticulation($beam);
			$fe = $this->getFermata($beam);
			$tu = $this->getTuplet($beam);
			$bb = $this->bb($beam->multiplicity);
			$bu = $this->bu($beam->up);
			$nr = $this->nr($beam->number);
			$sl = $this->nr($beam->slant);
			$ht = $this->ht($beam);
			$hta = $this->hta($beam);
			$pt = $this->pt($beam->long);
			$partial = $this->partial($beam->partial, $nr, $bu);
			$znotes = $this->getZBeams($beam, $nr);
			return $nl.$conn.$art.$fe.$tu.'\i'.$bb.$bu.$nr.$ht.$sl.$partial.$znotes.'\qb'.$pt.$nr.$hta;
		}

		protected function getBeamContinue(BeamContinue $beam) {
			$nl = $this->nl($beam->nolyr);
			$conn = $this->getConnections($beam);
			$art = $this->getArticulation($beam);
			$fe = $this->getFermata($beam);
			$bu = $this->bu($beam->up);
			$nr = $this->nr($beam->number);
			$change = $this->change($beam->change, $nr, $bu);
			$hta = $this->hta($beam);
			$pt = $this->pt($beam->long);
			$partial = $this->partial($beam->partial, $nr, $bu);
			$znotes = $this->getZBeams($beam, $nr);
			return $nl.$conn.$art.$fe.$change.$partial.$znotes.'\qb'.$pt.$nr.$hta;
		}

		protected function getBeamEnd(BeamEnd $beam) {
			$nl = $this->nl($beam->nolyr);
			$conn = $this->getConnections($beam);
			$art = $this->getArticulation($beam);
			$fe = $this->getFermata($beam);
			$bb = $this->bb($beam->multiplicity);
			$bu = $this->bu($beam->up);
			$nr = $this->nr($beam->number);
			$hta = $this->hta($beam);
			$pt = $this->pt($beam->long);
			$partial = $this->partial($beam->partial, $nr, $bu);
			$znotes = $this->getZBeams($beam, $nr);
			return $nl.$conn.$art.$fe.$partial.'\tb'.$bu.$nr.$znotes.'\qb'.$pt.$nr.$hta;
		}
		protected function getGchord(Chord $gchord=NULL) {
			if (is_null($gchord)) {
				return '';
			}
			if (!is_null($gchord->unsupported)) {
				$this->writeUnsupported($gchord->unsupported);
			}
			$text = $gchord->base;
			if ($gchord->alter === 1) {
				$text .= '^\#';
			} elseif ($gchord->alter === -1) {
				$text .= '^b';
			}
			if ($gchord->power) {
				$text .= '5';
			}
			if ($gchord->minor) {
				$text .= 'mi';
			}
			if ($gchord->sus > 0) {
				$text .= 'sus'.$gchord->sus;
			}
			if ($gchord->plus === 1) {
				$text .= '^{+}';
			} elseif ($gchord->plus === -1) {
				$text .= '^{-}';
			}
			if ($gchord->seven) {
				$text .= '7';
			}
			if ($gchord->maj) {
				$text .= '^{maj}';
			}
			return '\gchord{'.$text.'}';
		}

		protected function getConnections(Duration $note) {
			$connections = '';
			$ht = $this->ht($note);
			foreach ($note->connections as $connection) {
				$name = get_class($connection);
				$nr = $this->nr($connection->number);
				switch ($name) {
					case 'SlurStart':
						$cu = $this->cu($connection->up);
						$connections .= '\islur'.$cu.$nr.$ht;
						break;
					case 'SlurEnd':
						$connections .= '\tslur'.$nr.$ht;
						break;
					case 'TieStart':
						$cu = $this->cu($connection->up);
						$connections .= '\itie'.$cu.$nr.$ht;
						break;
					case 'TieEnd':
						$connections .= '\ttie'.$nr;
						break;
				}
			}
			foreach ($note->chords as $chord) {
				$connections .= $this->getConnections($chord);
			}
			return $connections;
		}

		protected function getFermata(Duration $duration) {
			if (!is_bool($duration->fermata)) {
				return '';
			}
			if ($duration->fermata) {
				$fermata = 'fermataup l';
			} else {
				$fermata = 'fermatadown f';
			}
			if ($duration instanceof Note) {
				$lng = $this->note[$duration->long];
				if (substr($lng, 1, 2) == 'wh') {
					$fermata = ucfirst($fermata);
				}
			}
			return '\\'.$fermata;
		}

		protected function getTuplet(Note $duration) {
			if (is_object($duration->tuplet)) {
				$this->writeUnsupported($duration->tuplet);
				return '';
			}
			if ($duration->tuplet == 3) {
				if ($duration instanceof Beam) {
					return '\triolet o';
				} else {
					return '\uptrio q20';
				}
			}
			if ($duration->tuplet == 2) {
				if ($duration instanceof Beam) {
					return '\xtuplet2o';
				} else {
					return '\def\txt{2}\uptrio q10';
				}
			}
			return '';
		}

		protected function getArticulation(Note $note) {
			$articulations = '';
			foreach ($note->articulations as $articulation) {
				$name = get_class($articulation);
				$articulations .= '\\'.($articulation->below ? 'l' : 'u');
				switch ($name) {
					case 'Accent':
						$articulations .= 'sf ';
						break;
					case 'Staccato':
						$articulations .= 'pz ';
						break;
				}
				$articulations .= $this->ht($note);
			}
			return $articulations;
		}

		protected function getZNotes(Note $note) {
			$result = '';
			foreach ($note->chords as $chord) {
				$lng = $this->znote[$chord->long];
				$hta = $this->hta($chord, ' ');
				$result .= $lng.$hta;
			}
			return $result;
		}

		protected function getZBeams(Beam $note, $nr) {
			$result = '';
			foreach ($note->chords as $chord) {
				$lng = '\zqb';
				$pt = $this->pt($chord->long);
				$hta = $this->hta($chord);
				$result .= $lng.$pt.$nr.$hta;
			}
			return $result;
		}

		protected function change($change, $nr, $bu) {
			if ($change > 0) {
				return '\tbb'.$bu.$nr;
			}
			if ($change < 0) {
				return '\nbb'.$bu.$nr;
			}
			return '';
		}

		protected function partial($partial, $nr, $bu) {
			if ($partial > 0) {
				return '\roff{\tbb'.$bu.$nr.'}';
			}
			if ($partial < 0) {
				return '\tbb'.$bu.$nr;
			}
			return '';
		}

		protected function ht(Note $note) {
			return $this->pitches[$note->height];
		}

		protected function hta(Note $note, $addit='') {
			$height = $this->ht($note);
			if (is_null($note->alter)) {
				return $addit.$height;
			}
			switch ($note->alter) {
				case -2:
					$height = '<'.$height;
					break;
				case -1:
					$height = '_'.$height;
					break;
				case 0:
					$height = '='.$height;
					break;
				case 1:
					$height = '^'.$height;
					break;
				case 2:
					$height = '>'.$height;
					break;
			}
			return '{'.$height.'}';
		}

		protected function nl($nolyr) {
			return $nolyr ? '\nolyr' : '';
		}

		protected function nr($number) {
			$number = (string) $number;
			return (strlen($number) > 1) ? '{'.$number.'}' : $number;
		}

		protected function bb($count) {
			return str_repeat('b', $count);
		}

		protected function bu($up) {
			return $up ? 'u' : 'l';
		}

		protected function cu($up) {
			return $up ? 'u' : 'd';
		}

		protected function pt($long) {
			return ($long % 3 == 0) ? 'p' : '';
		}

		protected function add($line) {
			$this->text .= $line."\n";
		}

	}



php-source/parser.php
<?php

	class Parser {

		private $pitches = array();

		private $types = array(
				'32nd'    => 1,
				'16th'    => 2,
				'eighth'  => 4,
				'quarter' => 8,
				'half'    => 16,
				'whole'   => 32,
				'breve'   => 64,
				'long'    => 128,
			);

		protected $measureNumber;

		protected $partNumber;

		public function __construct() {
			$names = array('C', 'D', 'E', 'F', 'G', 'A', 'B');
			$height = 1;
			for ($octave=2; $octave<=6; $octave++) {
				foreach ($names as $name) {
					$this->pitches[$octave][$name] = $height;
					$height++;
				}
			}
		}

		public function getMusic($xml) {
			$music = new Music;
			$multiple = 0;
			$x_measures = $xml->xpath('//measure');
			foreach ($x_measures as $x_measure) {
				$this->measureNumber = (string) $x_measure['number'];
				$measure = new Measure;
				$x_parts = $x_measure->xpath('part');
				foreach ($x_parts as $x_part) {
					$this->partNumber = (string) $x_part['id'];
					$part = new Part;

					$x_attributes = $x_part->xpath('attributes');
					if (!empty($x_attributes)) {
						assert('count($x_attributes)===1');
						$x_attributes = $x_attributes[0];
						foreach ($x_attributes as $x_attribute) {
							$name = $x_attribute->getName();
							switch ($name) {
								case 'time':
									$meter = new Meter;
									$meter->beats = (string) $x_attribute->beats;
									$meter->unit = (string) $x_attribute->{'beat-type'};
									$part->attributes[] = $meter;
									break;
								case 'key':
									$signature = new Signature;
									$signature->fifths = (string) $x_attribute->fifths;
									$part->attributes[] = $signature;
									break;
								case 'measure-style':
									if (isset($x_attribute->{'multiple-rest'})) {
										$multiple = (int) $x_attribute->{'multiple-rest'};
										$multiplePause = new FullPause;
										$multiplePause->long = $this->types['whole'];
										$multiplePause->count = $multiple;
									}
									break;
							}
						}
					}

					$x_notes = $x_part->xpath('(note|harmony)');
					$previous = NULL;
					$gchord = NULL;
					$tuplet = 0;
					foreach ($x_notes as $x_note) {
						if ($x_note->getName() === 'harmony') {
							$gchord = new Chord;
							assert('isset($x_note->root->{\'root-step\'})');
							$gchord->base = (string) $x_note->root->{'root-step'};
							if (isset($x_note->kind)) {
								$kind = (string) $x_note->kind;
								switch ($kind) {
									case 'major':
										break;
									case 'minor':
										$gchord->minor = TRUE;
										break;
									case 'dominant':
										$gchord->seven = TRUE;
										break;
									case 'minor-seventh':
										$gchord->minor = TRUE;
										$gchord->seven = TRUE;
										break;
									case 'augmented':
										$gchord->plus = 1;
										break;
									case 'diminished':
										$gchord->plus = -1;
										break;
									case 'major-seventh':
										$gchord->seven = TRUE;
										$gchord->maj = TRUE;
										break;
									case 'major-minor':
										$gchord->minor = TRUE;
										$gchord->seven = TRUE;
										$gchord->maj = TRUE;
										break;
									case 'suspended-second':
										$gchord->sus = 2;
										break;
									case 'suspended-fourth':
										$gchord->sus = 4;
										break;
									case 'power':
										assert('$x_note->kind["text"] == 5');
										$gchord->power = TRUE;
										break;
									default:
										$gchord->unsupported = $this->getUnsupported('HARMONY KIND', $gchord->base.' '.$kind);
								}
							}
							if (isset($x_note->root->{'root-alter'})) {
								$alter = (string) $x_note->root->{'root-alter'};
								switch ($alter) {
									case '1':
										$gchord->alter = 1;
										break;
									case '-1':
										$gchord->alter = -1;
										break;
									default:
										$gchord->unsupported = $this->getUnsupported('HARMONY ALTER', $gchord->base.' '.$alter);
								}
							}
						} else {
							if ($multiple !== 0) {
								assert('count($x_notes) === 1');
								$multiple--;
								if (empty($multiplePause)) {
									assert('isset($x_note->rest)');
									continue 3;
								}
								$duration = $multiplePause;
								unset($multiplePause);
							} else {
								$duration = $this->getDuration($x_note);
							}

							$x_ties = $x_note->xpath('notations/tied');
							foreach ($x_ties as $x_tie) {
								$type = (string) $x_tie['type'];
								if ($type == 'start') {
									$tie = new TieStart;
								} elseif ($type == 'stop') {
									$tie = new TieEnd;
								}
								$tie->number = (int) $x_tie['number'];
								$duration->connections[] = $tie;
							}

							$x_slurs = $x_note->xpath('notations/slur');
							foreach ($x_slurs as $x_slur) {
								$type = (string) $x_slur['type'];
								if ($type == 'start') {
									$slur = new SlurStart;
								} elseif ($type == 'stop') {
									$slur = new SlurEnd;
								}
								$slur->number = (int) $x_slur['number'];
								$duration->connections[] = $slur;
							}

							$x_tuplet = $x_note->xpath('time-modification');
							if (!empty($x_tuplet)) {
								assert('count($x_tuplet)===1');
								$x_tuplet = $x_tuplet[0];
								$actual = (int) $x_tuplet->{'actual-notes'};
								$normal = (int) $x_tuplet->{'normal-notes'};
								if (
									(($actual == 3) && ($normal == 2))
									||
									(($actual == 2) && ($normal == 3))
								) {
									if ($tuplet == 0) {
										$duration->tuplet = $actual;
										$tuplet = $actual - 1;
									} else {
										$tuplet--;
									}
								} else {
									$duration->tuplet = $this->getUnsupported('TUPLET', $actual.' for '.$normal);
								}
							} else {
								assert('$tuplet===0');
							}

							$x_fermatas = $x_note->xpath('notations/fermata');
							foreach ($x_fermatas as $x_fermata) {
								$type = (string) $x_fermata['type'];
								$duration->fermata = ($type != 'inverted');
							}

							$x_accents = $x_note->xpath('notations/articulations/accent');
							foreach ($x_accents as $x_accent) {
								assert('$duration instanceof Note');
								$placement = (string) $x_accent['placement'];
								$accent = new Accent;
								$accent->below = $placement == 'below';
								$duration->articulations[] = $accent;
							}

							$x_staccatos = $x_note->xpath('notations/articulations/staccato');
							foreach ($x_staccatos as $x_staccato) {
								assert('$duration instanceof Note');
								$placement = (string) $x_staccato['placement'];
								$staccato = new Staccato;
								$staccato->below = $placement == 'below';
								$duration->articulations[] = $staccato;
							}

							$x_chord = $x_note->xpath('chord');
							if (!empty($x_chord)) {
								assert('$previous instanceof Note');
								assert('get_class($duration)==="Note"');
								$previous->chords[] = $duration;
								continue;
							}

							$duration->gchord = $gchord;
							$gchord = NULL;

							$part->durations[] = $duration;
							$previous = $duration;
						}
					}

					$x_barlines = $x_part->xpath('barline');
					if (!empty($x_barlines)) {
						foreach ($x_barlines as $x_barline) {
							$x_ending = $x_barline->xpath('ending');
							if (!empty($x_ending)) {
								assert('count($x_ending)===1');
								$x_ending = $x_ending[0];
								$type = (string) $x_ending['type'];
								$volta = new Volta;
								$volta->number = (string) $x_ending['number'];
								if ($type == 'start') {
									$volta->start = TRUE;
								} elseif ($type == 'stop') {
									$volta->start = FALSE;
								} else {
									$volta = $this->getUnsupported('VOLTA TYPE', 'number: '.$volta->number.', type: '.$type);
								}
								$part->endings[] = $volta;
							}

							$x_repeat = $x_barline->xpath('repeat');
							if (!empty($x_repeat)) {
								assert('count($x_repeat)===1');
								$x_repeat = $x_repeat[0];
								$direction = (string) $x_repeat['direction'];
								$repeat = new Repeat;
								$repeat->left = ($direction == 'forward');
								$repeat->right = ($direction == 'backward');
								$part->endings[] = $repeat;
							}

							$x_double = $x_barline->xpath('bar-style');
							if (!empty($x_double)) {
								assert('count($x_double)===1');
								$x_double = $x_double[0];
								$style = (string) $x_double;
								if ($style == 'light-light') {
									$part->endings[] = new DoubleBar;
								}
							}
						}
					}
					$measure->parts[] = $part;
				}
				$music->measures[] = $measure;
			}
			return $music;
		}

		protected function getDuration($x_note) {
			$rest = FALSE;
			$pointed = 0;
			$alter = NULL;
			$accidental = NULL;
			$beams = array();
			foreach ($x_note as $info) {
				$name = $info->getName();
				switch ($name) {
					case 'pitch':
						$step = (string) $info->step;
						$octave = (string) $info->octave;
						if (!isset($this->pitches[$octave][$step])) {
							return $this->getUnsupported('NOTE PITCH', 'octave: '.$octave.', step: '.$step);
						}
						$height = $this->pitches[$octave][$step];
						if (isset($info->alter)) {
							$alter = (string) $info->alter;
							if (!in_array($alter, array('-2', '-1', '1', '2'))) {
								return $this->getUnsupported('PITCH ALTER', 'value: '.$alter);
							}
						}
						break;
					case 'accidental':
						$accidental = (string) $info;
						if (!in_array($accidental, array('double-sharp', 'sharp', 'natural'))) {
							return $this->getUnsupported('ACCIDENTAL', 'name: '.$accidental);
						}
						break;
					case 'type':
						$type = (string) $info;
						if (!isset($this->types[$type])) {
							return $this->getUnsupported('NOTE TYPE', 'type: '.$type);
						}
						$long = $this->types[$type];
						break;
					case 'beam':
						$beams[] = (string) $info;
						break;
					case 'rest':
						$rest = TRUE;
						break;
					case 'dot':
						$pointed++;
						break;
				}
			}

			if ($pointed > 1) {
				return $this->getUnsupported('MORE DOTS', 'dots: '.$pointed.', octave: '.$octave.', step: '.$step);
			}
			if ($pointed == 1) {
				if ($long == 1) {
					return $this->getUnsupported('DOT FOR 32nd', 'octave: '.$octave.', step: '.$step);
				}
				$long *= 1.5;
			}
			
			if (isset($height)) {
				if (!empty($beams)) {
					$duration = $this->getBeam($beams);
				} else {
					$duration = new Note();
				}
				$duration->height = $height;
				if ($accidental === 'natural') {
					$duration->alter = 0;
				} elseif (!\is_null($accidental)) {
					$duration->alter = (int) $alter;
				}
			} elseif ($rest) {
				$duration = new Pause;
			} else {
				return $this->getUnsupported('NOTE', 'unknown type');
			}
			$duration->long = $long;
			return $duration;
		}

		protected function getBeam(array $beams) {
			$begin = array();
			$continue = array();
			$end = array();
			$partial = 0;
			foreach ($beams as $beam) {
				switch ($beam) {
					case 'begin':
						$begin[] = $beam;
						break;
					case 'continue':
						$continue[] = $beam;
						break;
					case 'end':
						$end[] = $beam;
						break;
					case 'forward hook':
						assert('$partial == 0');
						$partial = 1;
						break;
					case 'backward hook':
						assert('$partial == 0');
						$partial = -1;
						break;
					default:
						return $this->getUnsupported('BEAM TYPE', 'type: '.$beam);
				}
			}
			$duration = NULL;
			if (!empty($continue)) {
				$duration = new BeamContinue();
				$duration->multiplicity = count($continue);
				$duration->multiplicity += count($begin);
			} elseif (!empty($begin)) {
				$duration = new BeamStart();
				$duration->multiplicity = count($begin);
				assert('count($end)==0');
			} elseif (!empty($end)) {
				$duration = new BeamEnd();
				$duration->multiplicity = count($end);
			}
			assert('!is_null($duration)');
			if ($duration->multiplicity > 3) {
				return $this->getUnsupported('BEAM MULTIPLICITY', 'count: '.$duration->multiplicity);
			}
			if ($partial == 1) {
				assert('$duration instanceof BeamStart || $duration instanceof BeamContinue');
			} elseif ($partial == -1) {
				assert('$duration instanceof BeamEnd || $duration instanceof BeamContinue');
			}
			$duration->partial = $partial;
			return $duration;
		}

		protected function getUnsupported($feature, $info) {
			$unsupported = new Unsupported;
			$unsupported->info = 'UNSUPPORTED '.$feature.' --  measure: '.$this->measureNumber.', part: '.$this->partNumber.', '.$info;
			return $unsupported;
		}

	}

